rclpy/rclpy/impl/rcutils_logger.py

```diff
@@ -0,0 +1,286 @@
+            )
+    return detected_filters
+
+
+class RcutilsLogger:
+
+    def __init__(self, name=''):
+        self.name = name
+        self.contexts = {}
+
+    def get_severity_threshold(self):
+        return _rclpy_logging.rclpy_logging_get_severity_threshold()
+
+    def set_severity_threshold(self, severity):
+        return _rclpy_logging.rclpy_logging_set_severity_threshold(severity)
+
+    def log(self, message, severity, **kwargs):
+        name = kwargs.pop('name', self.name)
+
+        # Infer the requested log filters from the keyword arguments
+        detected_filters = get_filters_from_kwargs(**kwargs)
+
+        # Get/prepare the context corresponding to the caller.
+        caller_id = CallerId()
+        caller_id_str = pickle.dumps(caller_id)
+        if caller_id_str not in self.contexts:
+            context = {'name': name, 'severity': severity}
+            for detected_filter in detected_filters:
+                if detected_filter in supported_filters:
+                    supported_filters[detected_filter].initialize_context(context, **kwargs)
+            context['filters'] = detected_filters
+            self.contexts[caller_id_str] = context
+        else:
+            context = self.contexts[caller_id_str]
+            # Don't support any changes to the logger.
+            if severity != context['severity']:
+                raise ValueError('Logger severity cannot be changed between calls.')
+            if name != context['name']:
+                raise ValueError('Logger name cannot be changed between calls.')
+            if detected_filters != context['filters']:
+                raise ValueError('Requested logging filters cannot be changed between calls.')
+            for detected_filter in detected_filters:
+                filter_params = supported_filters[detected_filter].params
+                if any(context[p] != kwargs.get(p, filter_params[p]) for p in filter_params):
+                    raise ValueError(
+                        'Logging filter parameters cannot be changed between calls.')
+
+        # Determine if it's appropriate to process the message (any filter can vote no)
+        # Note(dhood): even if a message doesn't get logged, a filter might still update its state
+        # as if it had been. This matches the behavior of the C logging macros provided by rcutils.
+        make_log_call = True
+        for logging_filter in context['filters']:
+            if not supported_filters[logging_filter].should_log(context):
+                make_log_call = False
+                break
+        # Only log the message if the severity is appropriate.
+        make_log_call &= severity >= self.get_severity_threshold()
+        if make_log_call:
+            # Call the relevant function from the C extension.
+            _rclpy_logging.rclpy_logging_rcutils_log(
+                severity, name, message,
+                caller_id.function_name, caller_id.file_name, caller_id.line_number)
```

Comment:
This could be refactored without the boolean variable to:

```
for logging_filter in context['filters']:
    if not supported_filters[logging_filter].should_log(context):
        break
else:
    if severity >= self.get_severity_threshold():
        # Call the relevant function from the C extension.
```

Reply:
yeah, I think the original reason for not doing so was to have a variable boolean to return. So to use the suggested structure we would need to add return statements in all condition branches, don't we?
```
for logging_filter in context['filters']:
    if not supported_filters[logging_filter].should_log(context):
        return False
else:
    if severity >= self.get_severity_threshold():
        # Call the relevant function from the C extension.
        return True
   else:
        return False
```

Reply:
Oh, I see. You would indeed need a `return True in the `if` block and one `return False` at the very end.

Reply:
This could even be adapted to follow the "return early" philosophy:

```
for logging_filter in context['filters']:
    if not supported_filters[logging_filter].should_log(context):
        return False

if severity < self.get_severity_threshold():
    return False

# Call the relevant function from the C extension.
...
return True
```

Reply:
thanks @mikaelarguedas and @dirk-thomas for the pointers! This is much cleaner now: e6a85ca2e4d9793264a1cbe03d595fc25c0e8560